### ISSUE TITLE
The "Dataset:Relation" function is not available in the admin page

### ISSUE THREAD

**Issue by YFan0019** [gigascience/gigadb-website#1959]
**Describe the bug**
The "Dataset:Relation" function is not available in the admin page

**To Reproduce**
1. Go to https://gigadb.org/adminRelation/admin
2. Click on "Create a new relation"
3. Try to enter the doi and relation, click create.
4. See error image

![GigaDB-Create-AdminRelation](https://github.com/gigascience/gigadb-website/assets/112144621/489957d9-7abc-4422-bcaf-0f96f2926478)

**Expected behavior**
The new relation should be saved in the system.

**Screenshots**

**Desktop (please complete the following information):**
 - OS:  iOS
 - Browser chrome
 - Version 126.0.6478.127

**Smartphone (please complete the following information):**

**Additional context**

**Comment by YFan0019** [comment]
Please add the relationship to the datasets from the backend server.
https://github.com/gigascience/dataset-tracking/issues/8

102538 HasPart 102546, 102545, 102544, 102543, 102542
https://gigadb.org/dataset/102538/token/h5r1bskBP6UORwni

102542 isPartOf 102538
https://gigadb.org/dataset/102542/token/79Vgzra-97MVu9s4

102543 isPartOf 102538
https://gigadb.org/dataset/102543/token/s7LBQcAmPo_cImVg

102544 isPartOf 102538
https://gigadb.org/dataset/102544/token/vEKNF93Qi945PzNo

102545 isPartOf 102538
https://gigadb.org/dataset/102545/token/hj-FXi-Ni7urZ7_A

102546 isPartOf 102538
https://gigadb.org/dataset/102546/token/A2QvodUl-z1TdXO_

**Comment by rija** [comment]
HI @YFan0019 

I have run the SQL query to create the relationships on Staging (https://staging.gigadb.org/). 
Can you check that it works for you. 
If yes, I will then run the query on the live database.

**Comment by YFan0019** [comment]
Hi @rija 

Looks good for me.
Thank you so much :)

**Comment by rija** [comment]
Hi @YFan0019,

I have now run the query on the live environment.
Let me know if it looks Ok.

For info the query is:
```
insert into relation (dataset_id, related_doi,relationship_id) values (2805,102546,12),(2805,102545,12),(2805,102544,12),(2805,102543,12),(2805,102542,12);

insert into relation (dataset_id, related_doi,relationship_id) values(2809,102538,11);
insert into relation (dataset_id, related_doi,relationship_id) values(2810,102538,11);
insert into relation (dataset_id, related_doi,relationship_id) values(2811,102538,11);
insert into relation (dataset_id, related_doi,relationship_id) values(2814,102538,11);
insert into relation (dataset_id, related_doi,relationship_id) values(2815,102538,11);
```

I'll live the ticket open, as we still need to fix the admin form for the functionality.

**Comment by YFan0019** [comment]
Hi @rija 
 
I checked the live server, and the datasets look all good.
I will release them in a minute.

**Comment by alli83** [comment]
Related to this ticket, what are the authorised names for relationships?

**Comment by only1chunts** [comment]
They should be in a table in the database, from memory I htink it was named something like "relationships", but they are also listed on the GigaDB help pages:
https://staging.gigadb.org/site/help
Scroll down to the section called "DOI relationship"

**Comment by only1chunts** [comment]
The DataCite schema from which that list was taken has been updated with a bunch more allowable relationships, so we may wish to consider allowing those: https://datacite-metadata-schema.readthedocs.io/en/4.5/appendices/appendix-1/relationType/

**Comment by alli83** [comment]
ok, thanks! @only1chunts

**Comment by alli83** [comment]
- IsPublishedIn
- IsVariantFormOf
- IsOrginalFormOf
- IsIdenticalTo
- IsDerivedFrom
- IsSourceOf

What are their reciprocal relations?

**Comment by rija** [comment]
> * IsPublishedIn

The functionality is to relate one dataset to another. That relationship doesn't feel relevant to that
> * IsVariantFormOf
> * IsOrginalFormOf


it looks like  those two are reciprocal of each other

> * IsIdenticalTo

it is its own recriprocal relationship

> * IsDerivedFrom
> * IsSourceOf

and those two are reciprocal of eachother

**Comment by alli83** [comment]
name in relationship table shouldn't be nullable right?

**Comment by alli83** [comment]
also what should we do if the second part of the relation (the reciprocal relation) fails but the first part( the relation created through the form) is saved?

**Comment by rija** [comment]
> name in relationship table shouldn't be nullable right?

correct

>also what should we do if the second part of the relation (the reciprocal relation) fails but the first part( the relation created through the form) is saved?

The first part is still valid metadata about a dataset so it should be preserved.

At first glance, it seems to me we should create a housekeeping process that we can use to detect and fix lingering one-sided relationships.
but i wonder what other approaches can be taken

**Comment by alli83** [comment]
@rija do you mean a cronjob/async process?

**Comment by rija** [comment]
> @rija do you mean a cronjob/async process?

yes, because we already needs (ticket is not implemented yet) a housekeeping process for various other tasks

**Comment by rija** [comment]
Hi @only1chunts,

Could you confirm what do you think we should do if some computer error occurs between saving the outgoing relationship and the saving of the reciprocal relationship.
There  are two options we've talked about: 

1) In the form, we save the outgoing relationship assignment to the database (and the dataset page of that origin dataset will point to the related one).  If the error is caught we can show it to the curator editing the form.
Otherwise we can make a script that we run regularly to detect and reports on one-sided relationships stored in the database.

2) In the form, we wrap the saving of relationships in a transaction so that if the  aforementioned failure scenario occurs, the entire transaction is aborted and no relationship are saved at all and an error can be reported to the curator editing the form.

**Comment by only1chunts** [comment]
Thanks for the options. As I cannot think of a scenario where the reciprocal link is not required they should be tightly linked and therefore option 2 sounds more appropriate. 
In the event of a failure we will need to investigate why and fix it on a case by case basis.
As a thought, would it be possible to add a tickbox (that by default is ticked) to say and add reciprocal relation. That way if there is a scenario in the future when we dont want the reciprocal it would be possible. It also gives an indication to the user that they do not need to go in and add the reciprocal separately.

**Comment by rija** [comment]
closing as fixed in pr #1959
